src/btree/bt_sync.c

```diff
@@ -114,43 +114,50 @@ __sync_ref_is_obsolete(WT_SESSION_IMPL *session, WT_REF *ref)
-    if (__wt_ref_is_root(ref))
+    if (__wt_ref_is_root(ref)) {
+        __wt_verbose(session, WT_VERB_CHECKPOINT_GC, "%p: skipping root page", (void *)ref);
         return (false);
+    }
 
     /* Ignore internal pages, these are taken care of during reconciliation. */
-    if (!__wt_ref_is_leaf(session, ref))
+    if (!__wt_ref_is_leaf(session, ref)) {
+        __wt_verbose(session, WT_VERB_CHECKPOINT_GC, "%p: skipping internal page with parent: %p",
+          (void *)ref, (void *)ref->home);
         return (false);
+    }
 
-    /* Check for the page obsolete, if the page is modified and reconciled. */
-    mod = (ref->page != NULL) ? (ref->page->modify) : NULL;
-    if (mod != NULL && mod->rec_result == WT_PM_REC_REPLACE)
-        return (__wt_txn_visible_all(
-          session, mod->mod_replace.newest_stop_txn, mod->mod_replace.newest_stop_ts));
-
+    obsolete = false;
     multi_newest_stop_ts = WT_TS_NONE;
     multi_newest_stop_txn = WT_TXN_NONE;
-    if (mod != NULL && mod->rec_result == WT_PM_REC_MULTIBLOCK) {
+    addr = ref->addr;
+    mod = (ref->page != NULL) ? (ref->page->modify) : NULL;
+
+    /* Check for the page obsolete, if the page is modified and reconciled. */
+    if (mod != NULL && mod->rec_result == WT_PM_REC_REPLACE)
+        obsolete = __wt_txn_visible_all(
+          session, mod->mod_replace.newest_stop_txn, mod->mod_replace.newest_stop_ts);
+    else if (mod != NULL && mod->rec_result == WT_PM_REC_MULTIBLOCK) {
         /* Calculate the max stop time pair by traversing all multi addresses. */
         for (multi = mod->mod_multi, i = 0; i < mod->mod_multi_entries; ++multi, ++i) {
             multi_newest_stop_ts = WT_MAX(multi_newest_stop_ts, multi->addr.newest_stop_ts);
             multi_newest_stop_txn = WT_MAX(multi_newest_stop_txn, multi->addr.newest_stop_txn);
         }
-        return (__wt_txn_visible_all(session, multi_newest_stop_txn, multi_newest_stop_ts));
-    }
-
-    /* Check if the page is obsolete using the page disk address. */
-    addr = ref->addr;
-    if (addr != NULL) {
+        obsolete = __wt_txn_visible_all(session, multi_newest_stop_txn, multi_newest_stop_ts);
+    } else if (addr != NULL) {
+        /* Check if the page is obsolete using the page disk address. */
         if (!__wt_off_page(ref->home, addr)) {
             __wt_cell_unpack(session, ref->home, (WT_CELL *)addr, &vpack);
-            return (__wt_txn_visible_all(session, vpack.newest_stop_txn, vpack.newest_stop_ts));
-        }
-        return (__wt_txn_visible_all(session, addr->newest_stop_txn, addr->newest_stop_ts));
+            obsolete = __wt_txn_visible_all(session, vpack.newest_stop_txn, vpack.newest_stop_ts);
+        } else
+            obsolete = __wt_txn_visible_all(session, addr->newest_stop_txn, addr->newest_stop_ts);
     }
 
-    return (false);
+    if (obsolete)
+        __wt_verbose(session, WT_VERB_CHECKPOINT_GC, "%p: page is found as obsolete", (void *)ref);
+    return (obsolete);
```

Comment:
The `__wt_off_page()` call will fail if `addr == NULL`, so I don't think you need both that call and a test for `addr == NULL`.

Reply:
Removed the unnecessary NULL check.